les/fetcher.go

```diff
@@ -625,28 +689,111 @@ func (f *lightFetcher) checkAnnouncedHeaders(fp *fetcherPeerInfo, headers []*typ
 // checkSyncedHeaders updates peer's block tree after synchronisation by marking
 // downloaded headers as known. If none of the announced headers are found after
 // syncing, the peer is dropped.
-func (f *lightFetcher) checkSyncedHeaders(p *peer) {
+func (f *lightFetcher) checkSyncedHeaders(p *peer) (bool, []*types.Header, *big.Int) {
 	fp := f.peers[p]
 	if fp == nil {
 		p.Log().Debug("Unknown peer to check sync headers")
-		return
+		return false, nil, nil
 	}
+
+	var h *types.Header
+	if f.pm.isULCEnabled() {
+		var unapprovedHashes []common.Hash
+		// Overwrite last announced for ULC mode
+		h, unapprovedHashes = f.lastTrustedTreeNode(p)
+		//rollback untrusted blocks
+		f.chain.Rollback(unapprovedHashes)
+	}
+
 	n := fp.lastAnnounced
 	var td *big.Int
+	trustedHeaderExisted := false
+
+	//find last trusted block
 	for n != nil {
-		if td = f.chain.GetTd(n.hash, n.number); td != nil {
+		//we found last trusted header
+		if n.hash == h.Hash() {
+			trustedHeaderExisted = true
+		}
+		if td = f.chain.GetTd(n.hash, n.number); td != nil && trustedHeaderExisted {
+			break
+		}
+
+		//break if we found last trusted hash before sync
+		if f.lastTrustedHeader == nil {
+			break
+		}
+		if n.hash == f.lastTrustedHeader.Hash() {
 			break
 		}
 		n = n.parent
 	}
-	// now n is the latest downloaded header after syncing
-	if n == nil {
+
+	// Now n is the latest downloaded/approved header after syncing
+	if n == nil && !p.isTrusted {
 		p.Log().Debug("Synchronisation failed")
 		go f.pm.removePeer(p.id)
-	} else {
-		header := f.chain.GetHeader(n.hash, n.number)
-		f.newHeaders([]*types.Header{header}, []*big.Int{td})
+		return false, nil, nil
 	}
+	header := f.chain.GetHeader(n.hash, n.number)
+	return true, []*types.Header{header}, td
+}
+
+// lastTrustedTreeNode return last approved treeNode and a list of unapproved hashes
+func (f *lightFetcher) lastTrustedTreeNode(p *peer) (*types.Header, []common.Hash) {
+	unapprovedHashes := make([]common.Hash, 0)
+	current := f.chain.CurrentHeader()
+
+	if f.lastTrustedHeader == nil {
+		return current, unapprovedHashes
+	}
+
+	canonical := f.chain.CurrentHeader()
+	if canonical.Number.Uint64() > f.lastTrustedHeader.Number.Uint64() {
+		canonical = f.chain.GetHeaderByNumber(f.lastTrustedHeader.Number.Uint64())
+	}
+	commonAncestor := rawdb.FindCommonAncestor(f.pm.chainDb, canonical, f.lastTrustedHeader)
+	if commonAncestor == nil {
+		log.Error("Common ancestor of last trusted header and canonical header is nil", "canonical hash", canonical.Hash(), "trusted hash", f.lastTrustedHeader.Hash())
+		return current, unapprovedHashes
+	}
+
+	for !f.isStopValidationTree(current, commonAncestor) {
```

Comment:
This condition checking function seems pretty pointless. We already know for sure that we have a commonAncestor and it is an ancestor of current. Checking for isTrustedHash again is also unnecessary. Checking for nil is enough to be absolutely safe, otherwise just iterate until you find the ancestor.

Reply:
I need it to get untrusted headers.